### ISSUE TITLE
4.0 Proposals

### ISSUE THREAD

**Issue by lweberk** [tomerfiliba-org/rpyc#106]
Hi, 

I've been looking at this library and I like a lot of features, specially the transparent approach to calling stuff in remote namespaces.
Though I would love to use the library the way it is, I cant help but think it can do even better;

1) its externalization approach quite invasive to be "pythonic"... I do not want to have to define methods with a determined prefix for them to be externalized, when I know about the existence of decorators :P...

``` python
 # class wise exposure (exposes whole class minus unsafe stuff which is defined by default, yet configurable like the "config" dict in the RPyC protocol.py)
@expose(alias='', namespace=None, lock=None)
class MyClass(object):

    # method wise exposure (analogously you can go for functions). Its the same decorator.
    @expose
    def my_sum(self, a, b): 
        return a + b 
```

The decorators optionally takes "alias", "namespace" , and "lock" as arguments;
- `alias`: Name that shows on the subscribers/importers/remote end.
- `namespace`: If you want to define a specific subset of all exposed methods/funcs and classes to be served at a specific serving reactor (more to this, so it makes sense in further points).
- `lock`: Instance of a Lock, that is acquired on call (making concurrency manageable, the same with namespace... will make more sense in further points)

2) I absolutely (deal killer) need a way to harness a n..m relationship. So far the approaches I have seen all try to build it "on top" of RPyC (please correct me if I am wrong). Now why not use the backend? But here lies a slight problem; it is extremely difficult to mess with the backend without breaking everything (networking is embedded up to the protocol). So basically you are restricted to use the provided backend, or resort to quite ugly hacks. Issue #84 
would solve part of it, but still it would be a long way for users to actually plug in their backend easily. I do not think it is an uncommon scenario that someone already has a server with its own event reactor, and would like to expose procedure calls through it using the RPyC protocol. Admittedly I'm one of those :D.

It would require;
- Complete segregation of networking from the actual protocol, the latter being only "called" and the synchronicity as well as the response or intermediate callbacks the sole responsibility of the networking side to route correctly. This way you could easily add another frame to the protocol stack, to have multiple clients calling at you, or inversely call at multiple clients giving you diverse responses (this last point actually takes us a little bit to the next point)
- Support for "multicall"; protocol has to make it possible to be explicitly instructed to return results in a list (if you expect multiple responses given a distributed backend) or in its original type (only 1..1). Either way its easy to implement, and has to be kept deterministic, programs have to be able to anticipate what they are getting. So no magic here.

Perhaps it sheds some light into my way of thought, if you picture ZMQ as malleable, clay-like backend. It can take pretty much any classic as well as unorthodox shape to serve, distribute, publish, subscribe etc.. to remote instances with RPyC exposed callables on their ends. Still it should not be directly supported by RPyC so not to incur further dependencies that are only justified if you actually want to build a backend with it.

3) Proposal for a central "Registry". Which keeps track and record of all exposed things, and over which locking and namespaces are enforced. Namespaces offer you a window/subset on the registered stuff, locks offer you a way of allowing concurrency with a way of making the exposed environment threadsafe. 
It also gives a nice entry point for zeroconf, since it would be possible to ask for the callables, even perhaps in a future for their argument types etc.
And finally it makes it also possible to interface with code written in other languages that are for instance statically typed. It would give way in a future to mask the input with type definitions on the python end, so stuff can be safely called from C/C++ for instance. This without additional stuff like swift or protobuffs define their callables call masks. (this is a little bit out of my realm, so please correct me if I am wrong)

4) The whole protocol only one class for both inbound as well as outbound calls. Though its "cleaner" if you get this approaches results served on a silver plate through a factory method, internally its hell for contributors and stakeholders. Proposing;
- RPyCProtocol: Gives you the interface to calls, wrapping and unwrapping etc (both children need this)
- RPyCExposer(RPyCProtocol): Interface to remotes. Its flow comes and is directed by the networking side (reactor), and given a callback for intermediate requests that might emanate in the work off process (not even this is necessary if you use UUID's, it would allow these calls to be made over the third class;)
- RPyCImporter(RPyCProtocol): Interface to remotely exposed and locally subscribed to classes/methods/functions. It is in itself nothing but the currently existing "proxy" object. Fully transparent. Its flow goes outward and is initiated by the program using the library.

5) Lets make RPyC 4.0 a fresh start...
- Only py3x support, remove all that compatibility clutter. Of course forks porting it down to py2x should be clearly encouraged, still keeping them as forks though.
- Remove all dev clutter as well :D. Yes, I did notice. Why not use feature branches? Git is perfect for this :P. And it keeps the master nicely pristine and free of unused code for those who want to contribute. Less confusion more contribution.
- Lets stick to PEP8. Use white lines where they should go etc. Yeah, I noticed that too. Any particular reason for that "block-gluing"?

Basically it would be a fresh release breaking back compatibility.

---

I know its a mouth full. And I also know it would take quite a bit of commitment. But hey... I am doing my best at finding a common ground for a better future and use of synergy, so please by all means tear this proposal to parts and bring your ideas into it, or if you like burn it all together.
As for the human power, I am committing myself and a co-worker at my company half-day for as long as it takes, if we get it to a stable release before the end of 2013.

Sincerely,
~lweberk
